AudioKit/Common/Nodes/Playback/Player/AKPlayer.swift

```diff
@@ -0,0 +1,724 @@
+    open var fade = Fade()
+
+    /// holds characteristics about the looping options
+    open var loop = Loop()
+
+    /// Volume 0.0 -> 1.0, default 1.0
+    open var volume: Float {
+        get { return playerNode.volume }
+        set { playerNode.volume = newValue }
+    }
+
+    /// Left/Right balance -1.0 -> 1.0, default 0.0
+    open var pan: Float {
+        get { return playerNode.pan }
+        set { playerNode.pan = newValue }
+    }
+
+    /// get or set the current time of the player. If playing it will scrub to that point.
+    /// If buffered it will preroll.
+    public var startTime: Double = 0 {
+        didSet {
+            startTime = max(0, startTime)
+
+            // Should playback restart automatically?
+            if isPlaying {
+                stop()
+                play()
+            }
+        }
+    }
+
+    public var endTime: Double = 0 {
+        didSet {
+            if endTime == 0 { endTime = duration }
+            endTime = min(endTime, duration)
+        }
+    }
+
+    /// - Returns: Current time of the player in seconds.
+    /// Use setTime() or startTime to set the time
+    open var currentTime: Double {
+        return position(at: nil)
+    }
+
+    // MARK: - public options
+
+    /// true if any fades have been set
+    open var isFaded: Bool {
+        return fade.inTime > 0 || fade.outTime > 0
+    }
+
+    /// true if the player is buffering audio rather than playing from disk
+    open var isBuffered: Bool {
+        if buffering == .never {
+            return false
+        } else {
+            return isReversed || isFaded || buffering == .always
+        }
+    }
+
+    open var isLooping: Bool = false
+
+    /// Reversing the audio will set the player to buffering
+    open var isReversed: Bool = false {
+        didSet {
+
+            let wasPlaying = isPlaying
+            stop()
+
+            if isLooping {
+                preroll(from: loop.start, to: loop.end)
+            }
+
+            if wasPlaying {
+                //play()
+            }
+        }
+    }
+
+    open var isPlaying: Bool {
+        return playerNode.isPlaying
+    }
+
+    // MARK: - Initialization
+
+    /// Create a player from a URL
+    public convenience init?(url: URL) {
+        if !FileManager.default.fileExists(atPath: url.path) {
+            return nil
+        }
+        do {
+            let avfile = try AVAudioFile(forReading: url)
+            self.init(audioFile: avfile)
+            return
+        } catch {
+            Swift.print(error)
+        }
+        return nil
+    }
+
+    /// Create a player from an AVAudioFile
+    public convenience init(audioFile: AVAudioFile) {
+        self.init()
+        self.audioFile = audioFile
+        initialize()
+    }
+
+    public override init() {
+        super.init(avAudioNode: mixer, attach: false)
+    }
+
+    private func initialize() {
+        guard audioFile != nil else { return }
```

Comment:
guard let audioFile = audioFile else { return }
Then get rid of the the forced unwrapping ```audioFile!```.

Reply:
ok will do. I have wondered about this convention. I find the syntax kind of more strange to create a new variable to unwrap then just guarding to make sure the original one isn't nil. But if this is the best convention, changing!

Reply:
is it in the case of if another thread could change audioFile ?

Reply:
It is exactly that.

Reply:
It's also just better to work with non optionals as a best practice.